Added AC systems

nloppi opened the pull request:
Hi Freddie,

There must be tons of stuff to modify. However, everything in this commit has been tested to be working.

Thanks,
Niki

FreddieWitherden commented on pyfr/plugins/fluidforce.py:
```diff
@@ -21,7 +21,10 @@ def __init__(self, intg, cfgsect, suffix):
 
         # Check if we need to compute viscous force
-        self._viscous = intg.system.name == 'navier-stokes'
+        self._viscous = intg.system.name == ('navier-stokes' or 'ac-navier-stokes')
```
This wont work.  `('navier-stokes' or 'ac-navier-stokes')` evaluates to `'navier-stokes'`.  The fluid force plugin is already messy enough with how it handles the viscous terms without yet another bunch of if statements to handle AC.  Rework.

FreddieWitherden commented on pyfr/plugins/fluidforce.py:
```diff
@@ -21,7 +21,10 @@ def __init__(self, intg, cfgsect, suffix):
+        self._viscous = intg.system.name == ('navier-stokes' or 'ac-navier-stokes')
+
+        # Check if the system is ac
+        self._ac = intg.system.name == ('ac-navier-stokes' or 'ac-euler')
```
See above for issues with this.

FreddieWitherden commented on pyfr/solvers/__init__.py:
```diff
@@ -4,6 +4,8 @@
 from pyfr.solvers.base import BaseSystem
 from pyfr.solvers.euler import EulerSystem
 from pyfr.solvers.navstokes import NavierStokesSystem
+from pyfr.solvers.aceuler import ACEulerSystem
```
Order imports alphabetically.

FreddieWitherden commented on pyfr/solvers/aceuler/elements.py:
```diff
@@ -0,0 +1,56 @@
+from pyfr.solvers.baseadvec import BaseAdvectionElements
+
+
+class BaseFluidElements(object):
```
Rename to `BaseACFluidElements` to avoid confusion with `BaseFluidElements` from Euler/Navier-Stokes.

FreddieWitherden commented on pyfr/solvers/aceuler/inters.py:
```diff
@@ -0,0 +1,83 @@
+        )
+
+        self._tpl_c.update(tplc)
+
```
Missing newline.

FreddieWitherden commented on pyfr/solvers/aceuler/kernels/bcs/ac-in-fv.mako:
```diff
@@ -0,0 +1,12 @@
+<%namespace module='pyfr.backends.base.makoutil' name='pyfr'/>
+
+<%pyfr:macro name='bc_rsolve_state' params='ul, nl, ur, ploc, t'>
+
```
Remove newline.

FreddieWitherden commented on pyfr/solvers/aceuler/kernels/bcs/ac-out-fp.mako:
```diff
@@ -0,0 +1,12 @@
+<%namespace module='pyfr.backends.base.makoutil' name='pyfr'/>
+
+<%pyfr:macro name='bc_rsolve_state' params='ul, nl, ur, ploc, t'>
+
```
Remove newline.

FreddieWitherden commented on pyfr/solvers/aceuler/kernels/bcs/ac-out-fp.mako:
```diff
@@ -0,0 +1,12 @@
+% for i, v in enumerate('uvw'[:ndims]):
+    ur[${i + 1}] = ul[${i + 1}];
+% endfor
+
```
Remove newline.

FreddieWitherden commented on pyfr/solvers/aceuler/kernels/bcs/ac-in-fv.mako:
```diff
@@ -0,0 +1,12 @@
+% for i, v in enumerate('uvw'[:ndims]):
+    ur[${i + 1}] = ${c[v]};
+% endfor
+
```
Remove newline.

FreddieWitherden commented on pyfr/solvers/aceuler/kernels/rsolvers/rusanov.mako:
```diff
@@ -0,0 +1,33 @@
+% endfor
+
+    // Normal of the average interface velocity
+    fpdtype_t nv = ${pyfr.dot('n[{i}]', 'vl[{i}] + vr[{i}]', i=ndims)}*0.5;
```
Move `0.5` to the front.

FreddieWitherden commented on pyfr/solvers/aceuler/kernels/rsolvers/rusanov.mako:
```diff
@@ -0,0 +1,33 @@
+    fpdtype_t a_1 = fabs(nv - sqrt(nv*nv+${c['ac-zeta']}));
+    fpdtype_t a_2 = fabs(nv + sqrt(nv*nv+${c['ac-zeta']}));
+    fpdtype_t a_3 = fabs(nv);
+    fpdtype_t a = max(max(a_1, a_2), a_3);
```
When can `a_1` be larger than `a_2`?

FreddieWitherden commented on pyfr/solvers/aceuler/system.py:
```diff
@@ -0,0 +1,15 @@
+# -*- coding: utf-8 -*-
+
+from pyfr.solvers.baseadvec import BaseAdvectionSystem
+from pyfr.solvers.aceuler.elements import ACEulerElements
```
Reorder imports alphabetically.

FreddieWitherden commented on pyfr/solvers/acnavstokes/inters.py:
```diff
@@ -0,0 +1,124 @@
+        super().__init__(*args, **kwargs)
+
+        self._tpl_c['v'] = self._eval_opts('uvw'[:self.ndims], default='0')
+
```
Missing newline.

FreddieWitherden commented on pyfr/solvers/acnavstokes/inters.py:
```diff
@@ -0,0 +1,124 @@
+class ACNavierStokesSlpWallBCInters(ACNavierStokesBaseBCInters):
+    type = 'slp-wall'
+    cflux_state = None
+
```
Missing newline.

FreddieWitherden commented on pyfr/solvers/acnavstokes/inters.py:
```diff
@@ -0,0 +1,124 @@
+        )
+
+        self._tpl_c.update(tplc)
+
```
Remove extra newline.

FreddieWitherden commented on pyfr/solvers/acnavstokes/inters.py:
```diff
@@ -0,0 +1,124 @@
+
+        # Pointwise template arguments
+        rsolver = self.cfg.get('solver-interfaces', 'riemann-solver')
+        tplargs = dict(ndims=self.ndims, nvars=self.nvars, rsolver=rsolver, c=self._tpl_c,
```
Line length.

FreddieWitherden commented on pyfr/solvers/acnavstokes/kernels/bcs/slp-wall.mako:
```diff
@@ -0,0 +1,30 @@
+% for i in range(nvars):
+    ul[${i}] = magnl*(ficomm[${i}]);
+% endfor
+</%pyfr:macro>
```
Missing newline.

FreddieWitherden commented on pyfr/solvers/acnavstokes/kernels/flux.mako:
```diff
@@ -0,0 +1,54 @@
+    fpdtype_t v_x = grad_uin[0][2];
+    fpdtype_t v_y = grad_uin[1][2];
+
+    fpdtype_t nu_c = ${c['nu']};
```
No need to define this.

FreddieWitherden commented on pyfr/solvers/acnavstokes/kernels/flux.mako:
```diff
@@ -0,0 +1,54 @@
+# -*- coding: utf-8 -*-
+<%namespace module='pyfr.backends.base.makoutil' name='pyfr'/>
+
+% if ndims == 2:
```
It should be easy to unify the `ndims = 2` and `ndims = 3` cases for this kernel (like we do for the inviscid flux).

FreddieWitherden commented on pyfr/solvers/acnavstokes/kernels/intcflux.mako:
```diff
@@ -0,0 +1,54 @@
+              gradur='in view fpdtype_t[${str(ndims)}][${str(nvars)}]'
+              nl='in fpdtype_t[${str(ndims)}]'
+              magnl='in fpdtype_t'>
+
```
Extra newline.

FreddieWitherden commented on pyfr/solvers/acnavstokes/elements.py:
```diff
@@ -0,0 +1,27 @@
+                u=self.scal_upts_inb, smats=self.smat_at('upts'),
+                f=self._vect_upts
+            )
+
```
Remove newline.

FreddieWitherden commented on pyfr/solvers/aceuler/system.py:
```diff
@@ -0,0 +1,15 @@
+from pyfr.solvers.baseadvec import BaseAdvectionSystem
+from pyfr.solvers.aceuler.elements import ACEulerElements
+from pyfr.solvers.aceuler.inters import (ACEulerIntInters, ACEulerMPIInters,
+                                       ACEulerBaseBCInters)
```
Indentation.

FreddieWitherden commented on pyfr/solvers/aceuler/kernels/flux.mako:
```diff
@@ -0,0 +1,24 @@
+<%pyfr:macro name='inviscid_flux' params='s, f'>
+    fpdtype_t v[${ndims}];
+
+    // Velocities directly in conservative variable array indices 1-3
```
Misleading comment in the case of `ndims = 2`.

FreddieWitherden commented on pyfr/solvers/acnavstokes/kernels/mpicflux.mako:
```diff
@@ -0,0 +1,52 @@
+    fvcomm += ${tau}*(ul[${i}] - ur[${i}]);
+% endif
+
+    ul[${i}] =  magnl*(ficomm[${i}] + fvcomm);
```
Extra space after = sign.

FreddieWitherden commented on pyfr/solvers/acnavstokes/kernels/mpiconu.mako:
```diff
@@ -0,0 +1,20 @@
+% endif
+% endfor
+</%pyfr:kernel>
+
```
Remove newline.

FreddieWitherden commented on pyfr/solvers/acnavstokes/kernels/intconu.mako:
```diff
@@ -0,0 +1,20 @@
+    ulout[${i}] = urout[${i}] = urin[${i}];
+% else:
+    ulout[${i}] = urout[${i}] = urin[${i}]*${0.5 + c['ldg-beta']}
+                        + ulin[${i}]*${0.5 - c['ldg-beta']};
```
Indentation.

FreddieWitherden commented on pyfr/solvers/aceuler/kernels/intcflux.mako:
```diff
@@ -0,0 +1,22 @@
+    ur[${i}] = -magnl*fn[${i}];
+% endfor
+</%pyfr:kernel>
+
```
Remove newline.

nloppi commented:
Hi Freddie,

The fluidforce plugin is still quite incovenient with plenty of code duplication. I will work on it, but first I'd like to have your opinion on the approach: Separate subclasses for ac and std systems that inherit from BaseFluidForcePlugin.

Thanks!

FreddieWitherden commented on pyfr/solvers/aceuler/kernels/rsolvers/rusanov.mako:
```diff
@@ -18,11 +18,8 @@
-    fpdtype_t a_3 = fabs(nv);
-    fpdtype_t a = max(max(a_1, a_2), a_3);
+    // Estimate the wave speed
+    fpdtype_t a = fabs(nv) + sqrt(nv*nv+${c['ac-zeta']});
```
Space around `+`.

FreddieWitherden commented on pyfr/plugins/fluidforce.py:
```diff
@@ -21,10 +19,7 @@ def __init__(self, intg, cfgsect, suffix):
-
-        # Check if the system is ac
-        self._ac = intg.system.name == ('ac-navier-stokes' or 'ac-euler')
+        self._viscous = ('navier-stokes' in intg.system.name)
```
Unnecessary parenthesis.

FreddieWitherden commented on pyfr/solvers/acnavstokes/kernels/flux.mako:
```diff
@@ -1,52 +1,18 @@
+% for i in range(ndims):
+    fout[${i}][${i+1}] += -2.0*${c['nu']}*(grad_uin[${i}][${i+1}]
+                        - ${1.0/3.0}*(${' + '.join('grad_uin[{j}][{k}]'
+                        .format(i=i,j=j,k=j+1) for j in range(ndims))}));
```
Various formatting issues.

FreddieWitherden commented on pyfr/solvers/acnavstokes/kernels/flux.mako:
```diff
@@ -1,52 +1,18 @@
+% endfor
+
+% for i in range(ndims-1):
+    fout[${i+1}][1] = fout[0][${i+2}] = -${c['nu']}*(grad_uin[${i+1}][1]
```
Various formatting issues.

FreddieWitherden commented on pyfr/solvers/acnavstokes/kernels/flux.mako:
```diff
@@ -1,52 +1,18 @@
+                        .format(i=i,j=j,k=j+1) for j in range(ndims))}));
+% endfor
+
+% for i in range(ndims-1):
```
Space around `-`.

FreddieWitherden commented on pyfr/plugins/fluidforce.py:
```diff
@@ -215,7 +208,88 @@ def stress_tensor(self, u, du):
 
-    def ac_stress_tensor(self, du):
+
+class ACFluidForcePlugin(BaseFluidForcePlugin):
```
Huge amounts of code duplication.  Move to base class.

FreddieWitherden commented on pyfr/solvers/acnavstokes/kernels/flux.mako:
```diff
@@ -0,0 +1,18 @@
+<%namespace module='pyfr.backends.base.makoutil' name='pyfr'/>
+<%pyfr:macro name='viscous_flux_add' params='uin, grad_uin, fout'>
+% for i in range(ndims):
+    fout[${i}][${i+1}] += -2.0*${c['nu']}*(grad_uin[${i}][${i+1}]
```
Move `-2.0` into `${}`.

FreddieWitherden commented on pyfr/solvers/acnavstokes/kernels/flux.mako:
```diff
@@ -0,0 +1,18 @@
+% for i in range(ndims):
+    fout[${i}][${i+1}] += -2.0*${c['nu']}*(grad_uin[${i}][${i+1}]
+                        - ${1.0/3.0}*(${' + '.join('grad_uin[{j}][{k}]'
+                        .format(i=i,j=j,k=j+1) for j in range(ndims))}));
```
Various formatting issues.

FreddieWitherden commented:
I think that inheritance is the correct approach to cleaning up the plugin.  However, currently, the design of the plugin infrastructure requires each class to have its own plugin name.  Hence, it is not possible for two classes to both provide a plugin called fluidforce necessitating one of them to be named `ac-fluidforce`.  I am unsure if I agree with this (and a similar principle applies to the naming of boundary conditions).  If so a mechanism will need to be developed for two classes to provide a plugin of the same name but for different systems.

nloppi commented:
The first expression in the viscous flux kernel still seems so messy. I tried to look reference from other kernels and found out that there is an oversized line in roe.mako. Has it just slipped through or is it sometimes legit to use longer expression in .mako files?

Thanks!

FreddieWitherden commented:
The first thing I would do is factor out the common part `u_x + v_y [+ w_z]` into its own variable.  This should help somewhat and then go from there.

nloppi commented:
I agree that having ac-fluidforce and fluidforce plugins is just confusing and I should work around that. However, I think that boundary conditions should be clearly separated as all AC bcs require different set of boundary values than their compressible counterparts except for the sub-out-fp. Moreover, terms like adiabatic and isothermal in the wall bcs do not apply to AC.

Thanks

FreddieWitherden commented on pyfr/solvers/aceuler/inters.py:
```diff
@@ -0,0 +1,85 @@
+        tplc, self._ploc = self._exp_opts(
+            ['p'], lhs
+        )
+
```
Remove newline.

FreddieWitherden commented on pyfr/solvers/acnavstokes/kernels/flux.mako:
```diff
@@ -0,0 +1,25 @@
+
+<%pyfr:macro name='viscous_flux_add' params='uin, grad_uin, fout'>
+    fpdtype_t divu = ${' + '.join('grad_uin[{i}][{j}]'
+                                  .format(i=i,j=i+1)
```
Space around `+`.

FreddieWitherden commented on pyfr/solvers/acnavstokes/kernels/flux.mako:
```diff
@@ -0,0 +1,25 @@
+
+% for i in range(ndims):
+    fout[${i}][${i + 1}] += ${-2.0*c['nu']}*(grad_uin[${i}][${i + 1}]
+                                             - ${1.0/3.0}*divu);
```
Move `1/3` up into the definition of `divu`.

FreddieWitherden commented on pyfr/solvers/acnavstokes/kernels/flux.mako:
```diff
@@ -0,0 +1,25 @@
+% endfor
+
+% for i in range(ndims - 1):
+    fpdtype_t t = ${-c['nu']}*(grad_uin[${i + 1}][1] + grad_uin[0][${i + 2}]);
```
Rename variable.

FreddieWitherden commented on pyfr/solvers/acnavstokes/kernels/flux.mako:
```diff
@@ -0,0 +1,25 @@
+    fout[0][${i + 2}] += t;
+% endfor
+
+% if ndims == 3:
```
This is still a big ugly.

FreddieWitherden commented on pyfr/solvers/acnavstokes/kernels/bcs/slp-wall.mako:
```diff
@@ -0,0 +1,30 @@
+    ${pyfr.expand('rsolve', 'ul', 'ur', 'nl', 'ficomm')};
+
+% for i in range(nvars):
+    ul[${i}] = magnl*(ficomm[${i}]);
```
Remove parenthesis.

FreddieWitherden commented on pyfr/solvers/acnavstokes/inters.py:
```diff
@@ -0,0 +1,123 @@
+
+        # Pointwise template arguments
+        rsolver = self.cfg.get('solver-interfaces', 'riemann-solver')
+        tplargs = dict(ndims=self.ndims, nvars=self.nvars, rsolver=rsolver, c=self._tpl_c)
```
Line length too long.

FreddieWitherden commented on pyfr/solvers/acnavstokes/elements.py:
```diff
@@ -0,0 +1,26 @@
+from pyfr.solvers.baseadvecdiff import BaseAdvectionDiffusionElements
+
+
+class ACNavierStokesElements(BaseACFluidElements, BaseAdvectionDiffusionElements):
```
Line length too long.

FreddieWitherden commented on pyfr/solvers/acnavstokes/inters.py:
```diff
@@ -0,0 +1,123 @@
+
+        # Pointwise template arguments
+        rsolver = self.cfg.get('solver-interfaces', 'riemann-solver')
+        tplargs = dict(ndims=self.ndims, nvars=self.nvars, rsolver=rsolver, c=self._tpl_c)
```
Line length too long.

nloppi commented:
Hi Freddie,
The flux kernel is still ugly, but I think that other approaches would be even uglier. Whether you decide to loop over the first column, row or upper/lower diagonal and copy the corresponding values inside the loop you are still left with two elements that need to be handled separately.

The other option would be to generate t_xy, t_xz (first column values) with one loop, t_xz, t_yz (third column) with another and then copy the second column values t_xz -> t_zx and t_yz -> t_zy. This is exactly the same as the first approach but the t_xz isn't copied in the first instance but generated with its own expression. I find this even more inconvenient. There wouldn't be an if statement, but a loop that would loop over range(ndims-2), which is probably even more confusing (than this message :p).

Thanks,
Niki

FreddieWitherden commented on pyfr/solvers/acnavstokes/kernels/intconu.mako:
```diff
@@ -0,0 +1,20 @@
+              urout='out view fpdtype_t[${str(nvars)}]'>
+% for i in range(nvars):
+% if c['ldg-beta'] == -0.5:
+    ulout[${i}] = urout[${i}] = ulin[${i}];
```
This will need updating, see revised copy in the Navier-Stokes solver.

FreddieWitherden commented on pyfr/solvers/acnavstokes/kernels/flux.mako:
```diff
@@ -0,0 +1,25 @@
+
+<%pyfr:macro name='viscous_flux_add' params='uin, grad_uin, fout'>
+    fpdtype_t divu = ${1.0/3.0}*(${' + '.join('grad_uin[{i}][{j}]'
+                                  .format(i=i,j=i + 1)
```
Identation, the ``.format`` should be part of the join.

FreddieWitherden commented on pyfr/solvers/acnavstokes/kernels/flux.mako:
```diff
@@ -0,0 +1,25 @@
+% endfor
+
+% if ndims == 3:
+    fpdtype_t tau = ${-c['nu']}*(grad_uin[1][3] + grad_uin[2][2]);
```
This may error on some backends as we're redefining ``tau``.

FreddieWitherden commented on pyfr/solvers/acnavstokes/kernels/bcs/slp-wall.mako:
```diff
@@ -0,0 +1,30 @@
+<%include file='pyfr.solvers.acnavstokes.kernels.flux'/>
+
+<%pyfr:macro name='bc_ldg_state' params='ul, nl, ur, ploc, t'>
+    fpdtype_t nor = ${' + '.join('ul[{1}]*nl[{0}]'.format(i, i + 1)
```
Reverse order of ``ul`` and ``nl`` so that it reads more naturally.

FreddieWitherden commented on pyfr/solvers/acnavstokes/inters.py:
```diff
@@ -0,0 +1,125 @@
+    def __init__(self, be, lhs, elemap, cfgsect, cfg):
+        super().__init__(be, lhs, elemap, cfgsect, cfg)
+
+        tplc, self._ploc = self._exp_opts(
```
This will need updating.

FreddieWitherden commented on pyfr/solvers/acnavstokes/inters.py:
```diff
@@ -0,0 +1,125 @@
+    def __init__(self, be, lhs, elemap, cfgsect, cfg):
+        super().__init__(be, lhs, elemap, cfgsect, cfg)
+
+        tplc, self._ploc = self._exp_opts(
```
This will need updating.

FreddieWitherden commented on pyfr/solvers/aceuler/inters.py:
```diff
@@ -0,0 +1,83 @@
+    def __init__(self, be, lhs, elemap, cfgsect, cfg):
+        super().__init__(be, lhs, elemap, cfgsect, cfg)
+
+        tplc, self._ploc = self._exp_opts(
```
This will need updating.

FreddieWitherden commented on pyfr/solvers/aceuler/inters.py:
```diff
@@ -0,0 +1,83 @@
+    def __init__(self, be, lhs, elemap, cfgsect, cfg):
+        super().__init__(be, lhs, elemap, cfgsect, cfg)
+
+        tplc, self._ploc = self._exp_opts(
```
This will need updating.

FreddieWitherden commented on pyfr/solvers/aceuler/kernels/flux.mako:
```diff
@@ -0,0 +1,24 @@
+    f[${i}][0] = ${c['ac-zeta']}*v[${i}];
+% endfor
+
+    // Momentum fluxes
```
Are these momenta or velocity?

FreddieWitherden commented on pyfr/solvers/aceuler/kernels/bcs/slp-wall.mako:
```diff
@@ -0,0 +1,12 @@
+<%namespace module='pyfr.backends.base.makoutil' name='pyfr'/>
+
+<%pyfr:macro name='bc_rsolve_state' params='ul, nl, ur, ploc, t'>
+    fpdtype_t nor = ${' + '.join('ul[{1}]*nl[{0}]'.format(i, i + 1)
```
Reverse order of ul and nl so that it reads more naturally.

FreddieWitherden commented on pyfr/solvers/aceuler/elements.py:
```diff
@@ -0,0 +1,56 @@
+
+
+class BaseACFluidElements(object):
+    formulations = ['std', 'dual']
```
AC only works with dual time stepping.

FreddieWitherden commented on pyfr/plugins/fluidforce.py:
```diff
@@ -8,9 +8,7 @@
-class FluidForcePlugin(BasePlugin):
-    name = 'fluidforce'
-    systems = ['euler', 'navier-stokes']
+class BaseFluidForcePlugin(BasePlugin):
```
Lets leave this plugin out of the first version of this pull request.  We can sort it out later.

FreddieWitherden commented on pyfr/solvers/aceuler/kernels/rsolvers/rusanov.mako:
```diff
@@ -0,0 +1,30 @@
+    ${pyfr.expand('inviscid_flux', 'ur', 'fr')};
+
+% for i in range(ndims):
+    vl[${i}] = ul[${i + 1}];
```
``pyfr.array`` may help you with initalising these critters and avoid the need for an explicit for loop.

nloppi commented:
Sorry, found a mistake in rk4. We don't need the extra bank

nloppi commented:
Fixed! Now the steppers are as you implemented them. I just changed the add_with_dts method. Which approach do you prefer? This, where you get the flexibility to defined the currsoln bank for source, which we do not need now, but might need in the future. Or the other, where the source term is computed only once and the added at each stage. Thanks

nloppi commented:
This approach also saves one matrix bank.

FreddieWitherden commented on pyfr/solvers/acnavstokes/kernels/flux.mako:
```diff
@@ -3,23 +3,23 @@
 
 % for i in range(ndims - 1):
-    fpdtype_t tau = ${-c['nu']}*(grad_uin[${i + 1}][1]
+    fpdtype_t ac_tau = ${-c['nu']}*(grad_uin[${i + 1}][1]
```
Still redefining tau.

nloppi commented:
Hi Freddie,

My commit history is messy. Is it ok, if I would open a completely new pull request for the AC systems? Do you prefer submitting it against your feature/acns branch or directly to vincentlab/develop? It should be already quite polished.

Thanks

FreddieWitherden commented:
Hi Niki,

On 25/10/2016 08:36, Niki Loppi wrote:
> My commit history is messy. Is it ok, if I would open a completely new
> pull request for the AC systems? Do you prefer submitting it against
> your feature/acns branch or directly to vincentlab/develop? It should be
> already quite polished.

Submit against my branch.  We can then iterate from there.  It expect it
will take a couple of weeks to get things cleaned up before we are in a
position to merge.

Regards, Freddie.
